realm/realm-library/src/main/java/io/realm/internal/SharedGroup.java

```diff
@@ -39,6 +39,10 @@
     private boolean activeTransaction;
     private final Context context;
 
+    // Default value is {@code false} indicating that {@link #waitForChange()} ends without user
+    // interruption. {@code true} means {@link #waitForChange()} is stopped by user.
+    private boolean isStoppedWaitingByUser = false;
```

Comment:
This is accessed from multiple threads. Looking at the code below it looks like you need to synchronize waitForChange and stopWaitForChange and make this volatile.